# Data Collection

## Data collection methods
- This notebook uses the Tweepy package to download tweets for specified accounts from the Twitter API. In order to use the API, you need your own bearer key, which serves as your authentication into the API.  If you are planning to use the API, put your bearer key into the variable 'bearer_key' in the constants section.
- It uses a manually created account list file (accounts.csv) which has all of the Twitter handles to download tweets from, the class to assign to each handle, and how many tweets to get download from each handle.  
- The accounts file is important to keep current in order to avoid downloading tweets from the same account multiple times as users of the Twitter API are limited to the number of tweets that can be pulled in a one month time period.  As such, when tweets for a particular handle are downloaded, the tweets are immediately appended to a tweets CSV file (tweet_list.csv) and the accounts file is updated to mark that handle as 'done'. 

## Install necessary uncommon packages 

In [8]:
# Tweet downloader
! pip install tweepy

## Imports

In [9]:
import os
from os.path import exists
import tweepy
import csv
import pandas as pd
import string
import json
from time import sleep

## Constants

In [21]:
# This is the account list that we'll be pulling tweets from.  
account_list_file = "data/accounts.csv"  

# This is the file that we'll save tweet data to.
tweet_list_file = 'data/tweet_list.csv'  

# This is the file we save the twitter bearer key in.  Should be in folder included in gitignore b/c it's private info.
twitter_key_file = "private/twitter_key.json"

# Gets the twitter key from the private folder (not uploaded to github, included in gitignore)
def get_twitter_key():
    with open(twitter_key_file) as f:
        data = json.load(f)
    return data['bearer_key']

# This the key needed to download from the API.  Get this from twitter key file.
bearer_key = get_twitter_key()

## Functions

### Account list management functions

These functions manage the list of accounts to pull tweets from.

In [22]:
# This downloads the account list file (the file that has the accounts we'll pull tweets from)
def get_account_list_from_file():
    
    # Get list of accounts from CSV file
    df_accounts = pd.read_csv(account_list_file)

    # Create a dataframe from the file contents
    df_accounts.columns = [n.strip() for n in df_accounts.columns]
    df_accounts['Count_Plan'] = df_accounts['Count_Plan'].astype(int)
    df_accounts['Count_Actual'] = df_accounts['Count_Actual'].astype(int)
    df_accounts['Done'] = df_accounts['Done'].astype(bool)
    
    # Return the dataframe
    return df_accounts

# This function saves the account list.  It's saved after each handle download.  
def save_account_list_to_file():
    account_list.to_csv(account_list_file, index=False)
    
# This is to reset the account list file.  Should rarely be used unless we want to restart the downloads.
def reset_account_list_done():
    account_list['Done'] = False
    account_list['Count_Actual'] = 0
    save_account_list_to_file()

### Tweet downloading and saving functions

These functions perform the actual downloading and saving 

In [23]:
# Function to request tweets from the twitter API for a specified handle, specified number of tweets, and add the specified class to it
# Return the list of tweets

def get_tweets(username, class_, number_of_tweets):
    # This is the key to use to download the tweets
   
    client = tweepy.Client(bearer_token=bearer_key)
    user_id = client.get_user(username=username).data.id

    # Uses the paginator to request as many tweets as we want (paginator makes it possible to download more than 100 at a time
    tweets = []
    for tweet in tweepy.Paginator(client.get_users_tweets, user_id, tweet_fields=['created_at', 'author_id'],expansions=[''], max_results=100, exclude=['replies']).flatten(limit=number_of_tweets):
        # Scrub the text of any non-readable characters
        text = "".join(i for i in tweet.text if i in string.printable)
        # Scrub the text of any newlines
        text = text.replace("\n", " ")
        # Put the tweet info into a new dictionary
        tweets.append({
            "user_name"  : str(username),
            'class'      : str(class_),
            "id"         : str(tweet.id),
            "text"       : str(text),
            "author_id"  : str(tweet.author_id),
            "created_at" : str(tweet.created_at)
        })
    return tweets

# Function to append newly downloaded tweets to file
def append_to_tweet_file(tweets):
    field_names = ['user_name','class','id','text','author_id', 'created_at']
    
    # if the tweet data file doesn't exist, we're starting from scratch.  Make the file and put the headers at the top. 
    if not os.path.exists(tweet_list_file):
        with open(tweet_list_file, 'a') as csv_file:
            writer = csv.writer(csv_file, quoting=csv.QUOTE_NONNUMERIC) 
            writer.writerow(field_names)
            
    # Append the new data to file
    with open(tweet_list_file, 'a') as csv_file:
        writer = csv.writer(csv_file, quoting=csv.QUOTE_NONNUMERIC) 
        for t in tweets:
            writer.writerow([t['user_name'], t['class'], t['id'], t['text'], t['author_id'], t['created_at']])

# Function to pull the next handle from the accounts file and 
def get_next_handle():
    # Find first handle with a False in 'Done' 
    next_account = 0
    total_accounts = len(account_list)
    count = 0
    
    # Loop through the account list to find the next one that doesn't say 'Done'.  This is the next handle to download.  
    for n in range(0, total_accounts):
        if account_list.loc[n, 'Done'] == False:
            # Found next handle to download.  Break the loop.
            next_account = n
            break

    
    # Double check we found a handle that doesn't say Done, and then get the tweets for that handle
    if account_list.loc[next_account, 'Done'] == False:
        handle_to_get = account_list.loc[next_account,'Twitter handle']
        class_assignment = account_list.loc[next_account,'Class']
        number_to_get = account_list.loc[next_account,'Count_Plan']
        # Print what we are downloading
        print(f"Requesting {next_account+1}/{total_accounts-1}: {handle_to_get}, {class_assignment}, {number_to_get} tweets.  ", end="")

        tweetlist = get_tweets(handle_to_get, class_assignment, number_to_get)
        count = len(tweetlist)
        if count > 0:
            # We've got tweets.  Mark it done in the accounts file and save it. 
            print(f"  Received: {count} tweets.")
            append_to_tweet_file(tweetlist)
            account_list.loc[next_account, "Done"] = True
            account_list.loc[next_account, "Count_Actual"] = count
            save_account_list_to_file()

    return count  

Test pulling of tweets to make sure it works

In [24]:
get_tweets("tombrady", "test", 5)

[{'user_name': 'tombrady',
  'class': 'test',
  'id': '1626668046931697664',
  'text': 'Looks like appropriate parade behavior to me  @PatrickMahomes https://t.co/T69dLoWxmS',
  'author_id': '1108050829393707008',
  'created_at': '2023-02-17 19:40:59+00:00'},
 {'user_name': 'tombrady',
  'class': 'test',
  'id': '1626265734614589445',
  'text': 'Hey @Autograph family, were one month away from my unofficial retirement party in Tampa on March 16th. Cant wait to celebrate with everyone from The Huddle.   You down @RobGronkowski? Going over the list now and think I can get you in',
  'author_id': '1108050829393707008',
  'created_at': '2023-02-16 17:02:20+00:00'},
 {'user_name': 'tombrady',
  'class': 'test',
  'id': '1624910619559227392',
  'text': 'Two of the best tonight in Pat and Jalen. Good luck to KC and Philly, cant wait to watch.',
  'author_id': '1108050829393707008',
  'created_at': '2023-02-12 23:17:35+00:00'},
 {'user_name': 'tombrady',
  'class': 'test',
  'id': '162490855973

## Download tweets

Two download methods are provided below.  One to download one handle's tweets (the next handle in the account list that isn't downloaded yet).  One to download a batch of the next 50 handles in the account list.  

### Download the next handle's tweets

In [25]:
# Get account list from file
account_list = get_account_list_from_file()

# Download and save the next handle
count = get_next_handle()

### Download the next 50 handles' tweets

In [60]:
# Loop through the next 50 handles to pull from the account file
for n in range(50):
    count = get_next_handle()  # Returns the number of tweets downloaded.  If zero, end, something didn't work.  
    if count == 0:
        break
    # Sleep for 1 second and then move on to the next handle.  Give it time to download.
    sleep(1)  

## Data review

Review downloaded data and the account status file

### Review tweet data

Load downloaded tweets from file (assumes the tweet file already has downloaded tweets in it)

In [26]:
tweet_list_df = pd.read_csv(tweet_list_file)
tweet_list_df

user_name                 class            id  \
0       BennieGThompson    Politics - Liberal  1.620580e+18   
1       BennieGThompson    Politics - Liberal  1.620120e+18   
2       BennieGThompson    Politics - Liberal  1.620120e+18   
3       BennieGThompson    Politics - Liberal  1.620120e+18   
4       BennieGThompson    Politics - Liberal  1.620060e+18   
...                 ...                   ...           ...   
136667        LinusTech  Science / Technology  1.550000e+18   
136668        LinusTech  Science / Technology  1.550000e+18   
136669        LinusTech  Science / Technology  1.550000e+18   
136670        LinusTech  Science / Technology  1.550000e+18   
136671        LinusTech  Science / Technology  1.550000e+18   

                                                     text    author_id  \
0       Today marks the 83rd anniversary of the first ...   82453460.0   
1       RT @VP: President Biden and I are just getting...   82453460.0   
2       RT @RepJeffries: We will never negotiate away ...   82453460.0   
3                                 https://t.co/Ze7ePCUJJ2   82453460.0   
4         https://t.co/ley5hNsz0y https://t.co/RFdTeGXGO1   82453460.0   
...                                                   ...          ...   
136667  NEW VIDEO!: who would PAY for this Weird Fan? ...  403614288.0   
136668  Big shoutout to @LongMcQuade for providing us ...  403614288.0   
136669  New Video!!: Should I take back my $225,000?  ...  403614288.0   
136670  we finally got colton a standing desk https://...  403614288.0   
136671     upcoming DLC for GTA 6 https://t.co/jluyfFA5f9  403614288.0   

                       created_at  
0       2023-02-01 00:45:11+00:00  
1       2023-01-30 17:46:29+00:00  
2       2023-01-30 17:46:12+00:00  
3       2023-01-30 17:45:55+00:00  
4       2023-01-30 14:10:33+00:00  
...                           ...  
136667  2022-08-02 17:18:49+00:00  
136668  2022-08-02 17:13:55+00:00  
136669  2022-08-01 18:45:12+00:00  
136670  2022-08-01 16:02:07+00:00  
136671  2022-08-01 00:00:26+00:00  

[136672 rows x 6 columns]

In [27]:
tweet_list_df.user_name.value_counts()

WSJbusiness        1606
LinusTech           850
techreview          850
ScienceMagazine     850
NatGeo              850
                   ... 
repvalhoyle          10
JMoylanforGuam       10
NBCNetwork            7
RepJeffJackson        2
DNC                   1
Name: user_name, Length: 612, dtype: int64

It looks like we've successfully successfully downloaded from all accounts.

Notes:
- Tweets totaled ~136K in count from 612 accounts.  

Now I'll move on to modeling.  Proceed back to the main notebook. 